File: ad_library/src/main/java/com/adgrowth/internal/helpers/AdUriHelpers.java
package com.adgrowth.internal.helpers;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;

import com.adgrowth.adserver.AdServer;

import java.util.UUID;

public class AdUriHelpers {
    public static void openUrl(Activity context, String uri, String ipAddress) {

        uri = replaceAdCallbackParams(context, uri, ipAddress);

        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(uri));
        context.startActivity(intent);

    }

    public static String replaceAdCallbackParams(Activity context, String uri, String ipAddress) {
        String clickId = UUID.randomUUID().toString();
        String adId = AdServer.getAdId();
        String siteId = context.getPackageName();

        return uri
                .replaceAll("\\{advertising_id\\}", adId)
                .replaceAll("\\{click_id\\}", clickId)
                .replaceAll("\\{site_id\\}", siteId)
                .replaceAll("\\{ip\\}", ipAddress);
    }
}

File: ad_library/src/main/kotlin/com/adgrowth/internal/integrations/adserver/helpers/AdUriHelpers.kt
package com.adgrowth.internal.integrations.adserver.helpers

import android.app.Activity
import android.content.Intent
import android.net.Uri
import com.adgrowth.internal.integrations.InitializationManager
import java.util.*

object AdUriHelpers {
    @JvmStatic
    fun openUrl(context: Activity, url: String, ipAddress: String, uniqueId: String) {
        val intent =
            Intent(Intent.ACTION_VIEW, Uri.parse(replaceURISnippets(url, ipAddress, uniqueId)))
        context.startActivity(intent)
    }

    @JvmStatic
    fun replaceURISnippets(uri: String, ipAddress: String, uniqueId: String): String {
        val adId = InitializationManager.ADVERTISING_ID
        val appId = InitializationManager.APP_META_DATA.appId

        return uri
            .replace("\\{advertising_id\\}".toRegex(), adId)
            .replace("\\{click_id\\}".toRegex(), uniqueId)
            .replace("\\{site_id\\}".toRegex(), appId)
            .replace("\\{ip\\}".toRegex(), ipAddress)
    }

    fun getUniqueId(): String {
        return UUID.randomUUID().toString()
    }
}
